pennylane/devices/default_mixed.py

```diff
@@ -0,0 +1,276 @@
+
+        Args:
+            index (int): integer representing the computational basis state.
+
+        Returns:
+            array[complex]: complex array of shape ``[2] * (2 * num_wires)``
+            representing the density matrix of the basis state.
+        """
+        rho = np.zeros((2 ** self.num_wires, 2 ** self.num_wires), dtype=np.complex128)
+        rho[index, index] = 1
+        rho = self._asarray(rho, dtype=self.C_DTYPE)
+        return self._reshape(rho, [2] * (2 * self.num_wires))
+
+    @property
+    def state(self):
+        """Returns the state density matrix of the circuit prior to measurement"""
+        dim = 2 ** self.num_wires
+        # User obtains state as a matrix
+        return self._reshape(self._pre_rotated_state, (dim, dim))
+
+    def reset(self):
+        """Resets the device"""
+        super().reset()
+
+        self._state = self._create_basis_state(0)
+        self._pre_rotated_state = self._state
+
+    def analytic_probability(self, wires=None):
+
+        if self._state is None:
+            return None
+
+        # convert rho from tensor to matrix
+        rho = self._reshape(self._state, (2 ** self.num_wires, 2 ** self.num_wires))
+        # probs are diagonal elements
+        probs = self.marginal_prob(self._diag(rho), wires)
+        return probs
+
+    def _get_kraus(self, operation):  # pylint: disable=no-self-use
+        """Return the Kraus operators representing the operation.
+
+        Args:
+            operation (~.Operation): a PennyLane operation
+
+        Returns:
+            list[array[complex]]: Returns a list of 2D matrices representing the Kraus operators. If
+            the operation is unitary, returns a single Kraus operator. In the case of a diagonal
+            unitary, returns a 1D array representing the matrix diagonal.
+        """
+        if isinstance(operation, DiagonalOperation):
+            return operation.eigvals
+
+        if isinstance(operation, Channel):
+            return operation.kraus_matrices
+
+        return [operation.matrix]
+
+    def _apply_channel(self, kraus, wires):
+        r"""Apply a quantum channel specified by a list of Kraus operators to subsystems of the
+        quantum state. For a unitary gate, there is a single Kraus operator.
+
+        Args:
+            matrices (array): Kraus operators
+            wires (Wires): target wires
+        """
+
+        channel_wires = self.map_wires(wires)
+
+        # Computes K^\dagger , which are needed for the transformation K \rho K^\dagger
+        kraus_dagger = [self._conj(self._transpose(k)) for k in kraus]
+
+        # Changes tensor shape
+        kraus_shape = len(kraus) + [2] * len(device_wires) * 2
+        kraus = self._cast(self._reshape(kraus, kraus_shape), dtype=self.C_DTYPE)
+        kraus_dagger = self._cast(self._reshape(kraus_dagger, kraus_shape), dtype=self.C_DTYPE)
+
+        # Tensor indices of the state. For each qubit, need an index for rows *and* columns
```

Comment:
Something to note: both `ABC` and `np.einsum` support a maximum of 52 letters, which for the mixed state simulator means that we can simulate at most 26 qubits, since we need two indices per qubit to express the density matrix tensor. We actually need a few more indices, so maybe the upper limit is closer to 23 qubits

Reply:
It's likely the case that simulating ~20 qubits is infeasible on a typical desktop, so this may not be an issue. In `default.qubit` there is the same problem, but because the state has fewer indices, the limit is 26*2=52 qubits. There, `einsum` and `tensordot` are used depending on the number of wires the *operation* acts on, but you still need many indices to describe the state tensor.

Eventually, it may be good to write a version of `apply()` based on `tensordot`and benchmark performance of both approaches. For now, my strategy is to build a first working version of the device, then make an effort to optimize performance further, lightning.qubit style :zap:

Reply:
What happens if the number of qubits goes too high? Maybe we should place a limit in the `__init__` by raising an error if too many wires are requested? (As you say, this could be a shorter term fix)

Reply:
I just tried to do this and we anyway get a memory error if we want to initialize a device with 23 qubits 
![Screenshot from 2020-09-22 15-30-16](https://user-images.githubusercontent.com/40441298/93928666-ad665b00-fce8-11ea-840f-0bf56c16c0d8.png)